src/sign/records.rs

```diff
@@ -74,34 +77,122 @@ impl<N, D> SortedRecords<N, D> {
         self.rrsets().find(|rrset| rrset.rtype() == Rtype::SOA)
     }
 
+    /// Sign a zone using the given keys.
+    ///
+    /// A DNSKEY RR will be output for each key.
+    ///
+    /// Keys with a supported algorithm with the ZONE flag set will be used as
+    /// ZSKs.
+    ///
+    /// Keys with a supported algorithm with the ZONE flag AND the SEP flag
+    /// set will be used as KSKs.
+    ///
+    /// If only one key has a supported algorithm and has the ZONE flag set
+    /// AND has the SEP flag set, it will be used as a CSK (i.e. both KSK and
+    /// ZSK).
     #[allow(clippy::type_complexity)]
-    pub fn sign<Octets, Key, ApexName>(
+    pub fn sign<Octets, SigningKey>(
         &self,
-        apex: &FamilyName<ApexName>,
+        apex: &FamilyName<N>,
         expiration: Timestamp,
         inception: Timestamp,
-        key: Key,
-    ) -> Result<Vec<Record<N, Rrsig<Octets, ApexName>>>, Key::Error>
+        keys: &[(SigningKey, validate::Key<Octets>)], // private, public key pair
+    ) -> Result<Vec<Record<N, ZoneRecordData<Octets, N>>>, ()>
     where
         N: ToName + Clone,
-        D: RecordData + ComposeRecordData,
-        Key: SigningKey,
-        Octets: From<Key::Signature> + AsRef<[u8]>,
-        ApexName: ToName + Clone,
+        D: RecordData + ComposeRecordData + From<Dnskey<Octets>>,
+        SigningKey: SignRaw,
+        Octets: AsRef<[u8]>
+            + Clone
+            + From<Box<[u8]>>
+            + octseq::OctetsFrom<std::vec::Vec<u8>>,
     {
-        let mut res = Vec::new();
-        let mut buf = Vec::new();
+        // Per RFC 8624 section 3.1 "DNSSEC Signing" column guidance.
+        let unsupported_algorithms = [
+            SecAlg::RSAMD5,
+            SecAlg::DSA,
+            SecAlg::DSA_NSEC3_SHA1,
+            SecAlg::ECC_GOST,
+        ];
+
+        let mut ksks: Vec<&(SigningKey, validate::Key<Octets>)> = keys
+            .iter()
+            .filter(|(k, _)| !unsupported_algorithms.contains(&k.algorithm()))
+            .filter(|(_, dk)| {
+                dk.is_zone_signing_key() && dk.is_secure_entry_point()
+            })
+            .collect();
+
+        let mut zsks: Vec<&(SigningKey, validate::Key<Octets>)> = keys
+            .iter()
+            .filter(|(k, _)| !unsupported_algorithms.contains(&k.algorithm()))
+            .filter(|(_, dk)| {
+                dk.is_zone_signing_key() && !dk.is_secure_entry_point()
+            })
+            .collect();
+
+        // CSK?
+        if !ksks.is_empty() && zsks.is_empty() {
+            zsks = ksks.clone();
+        } else if ksks.is_empty() && !zsks.is_empty() {
+            ksks = zsks.clone();
+        }
```

Comment:
What happens if there are no ZSKs or KSKs?  Shouldn't the function stop entirely?